README.txt
==================================
Python ProfitBricks syslog modules
==================================

These are Python modules to extend the logging mechanism in Python.

Author: Frank Brehm (<frank.brehm@profitbricks.com>)



Module pb_logging.colored.ColoredFormatter
==========================================

Synopsis:
---------

An additional logging formatter for colored output via console.

Usage:
------

import logging
from pb_logging.colored import ColoredFormatter

logger = logging.getLogger(__name__)

format_str = '%(name)s: %(message)s (%(filename)s:%(lineno)d)'
formatter = ColoredFormatter(format_str)

console = logging.StreamHandler()
console.setFormatter(formatter)
logger.addHandler(console)

logger.debug('Debug')
logger.info('Info')
logger.warning('Warning!')
logger.error('ERROR!')
logger.critical('CRITICAL!!!')



Module pb_logging.syslog_handler.PbSysLogHandler
================================================

Synopsis:
---------

A wrapping logging handler for logging.handlers.SysLogHandler
to avoid BOM errors in syslog messages.

Usage:
------

import logging
from pb_logging.syslog_handler import PbSysLogHandler

logger = logging.getLogger(__name__)

appname = os.path.basename(sys.argv[0])
format_str = appname + ': %(name)s(%(lineno)d) %(funcName)s() ' + '%(levelname)s - %(message)s')
formatter = logging.Formatter(format_str)

lh_syslog = PbSysLogHandler(
        address = '/dev/log',
        facility = logging.handlers.SysLogHandler.LOG_USER,
)
lh_syslog.setFormatter(formatter)
logger.addHandler(lh_syslog)

logger.debug(u'An unicode message.')



Module pb_logging.unix_handler.UnixSyslogHandler
================================================

Synopsis:
---------

An additional logging handler for the common logging framework
to combine it with syslog.

Usage:
------

import logging
from pb_logging.unix_handler import UnixSyslogHandler

logger = logging.getLogger(__name__)

appname = os.path.basename(sys.argv[0])
format_str = appname + ': %(name)s(%(lineno)d) %(funcName)s() ' + '%(levelname)s - %(message)s')
formatter = logging.Formatter(format_str)

lh_syslog = UnixSyslogHandler(
        ident = appname,
        facility = logging.handlers.SysLogHandler.LOG_USER,
)
lh_syslog.setFormatter(formatter)
logger.addHandler(lh_syslog)

logger.debug('Bli bla blub.')



debian/changelog
pb-logging (0.5.5) all; urgency=low

  * Minor changes for newer Python3

 -- Frank Brehm <frank@brehm-online.com>  Tue, 12 Jun 2018 21:47:11 +0200

pb-logging (0.5.4) all; urgency=low

  * Applying Flake8 rules

 -- Frank Brehm <frank@lena.home.brehm-online.com>  Thu, 16 Feb 2017 11:43:51 +0100

pb-logging (0.5.3) all; urgency=low

  * Pointless version dump for building packages for Ubuntu zesty

 -- Frank Brehm <frank.brehm@profitbricks.com>  Tue, 22 Nov 2016 11:05:27 +0100

pb-logging (0.5.2) all; urgency=medium

  * Changing setup.py from distutils to setuptools.

 -- Frank Brehm <frank.brehm@profitbricks.com>  Tue, 14 Jun 2016 12:24:31 +0200

pb-logging (0.5.1) all; urgency=medium

  * Adding boolean parameter 'dark' to __init__() of ColoredFormatter to
    initialize it with dark colors instead of bold ones.
  * Adding two new unit test methods to test/test_colored.py

 -- Frank Brehm <frank.brehm@profitbricks.com>  Fri, 20 May 2016 11:33:15 +0200

pb-logging (0.4.11) all; urgency=medium

  * Changes on setup.py for moved Git URL
  * Apply flake8 rules to all modules, cosmetic changes
  * low Changed all copyrights from 2015 to 2016
  * Removed increase-version.sh and pb-logging.ebuild

 -- Frank Brehm <frank.brehm@profitbricks.com>  Tue, 19 Jan 2016 12:13:50 +0100

pb-logging (0.4.9) all; urgency=low

  * Added dh_python to build dependencies in debian/control.

 -- Frank Brehm <frank.brehm@profitbricks.com>  Fri, 03 Jul 2015 15:42:48 +0200

pb-logging (0.4.8) all; urgency=low

  * Another version bump for testing package build in new build chains

 -- Frank Brehm <frank.brehm@profitbricks.com>  Mon, 01 Jun 2015 13:05:46 +0200

pb-logging (0.4.7) all; urgency=low

  * Version bump for testing package build in new build chains

 -- Frank Brehm <frank.brehm@profitbricks.com>  Thu, 28 May 2015 13:17:03 +0200

pb-logging (0.4.6) all; urgency=medium

  * Switch to dpkg-source 3.0 (native) format
  * Update homepage and git URLs to point to gitlab

 -- Benjamin Drung <benjamin.drung@profitbricks.com>  Tue, 02 Dec 2014 15:42:04 +0100

pb-logging (0.4.5-4) squezze wheezy jessie; urgency=low

  * Release for different distros

 -- Frank Brehm <frank.brehm@profitbricks.com>  Wed, 12 Nov 2014 15:39:55 +0100

pb-logging (0.4.5-3) stable; urgency=low

  * Build for squeeze

 -- Frank Brehm <frank.brehm@profitbricks.com>  Mon, 27 Oct 2014 17:56:35 +0100

pb-logging (0.4.5-2) jessie; urgency=low

  * Build for jessie

 -- Frank Brehm <frank.brehm@profitbricks.com>  Mon, 27 Oct 2014 17:46:16 +0100

pb-logging (0.4.5-1) wheezy; urgency=low

  * Declared for unstable
  * Applied PEP8 rules to different files

 -- Frank Brehm <frank.brehm@profitbricks.com>  Sat, 25 Oct 2014 00:30:47 +0200

pb-logging (0.4.4-2) stable; urgency=low

  * Build for squeeze

 -- Frank Brehm <frank.brehm@profitbricks.com>  Wed, 24 Sep 2014 17:12:00 +0200

pb-logging (0.4.4-1) wheezy; urgency=low

  * Bugfixing and documentation in pb_logging/colored.py

 -- Frank Brehm <frank.brehm@profitbricks.com>  Wed, 24 Sep 2014 17:05:58 +0200

pb-logging (0.4.3-1) unstable; urgency=low

  * Declared for unstable

 -- Frank Brehm <frank.brehm@profitbricks.com>  Wed, 09 Jul 2014 14:02:13 +0200

pb-logging (0.4.2-4) stable; urgency=low

  * Build for squeeze

 -- Frank Brehm <frank.brehm@profitbricks.com>  Fri, 04 Jul 2014 13:05:22 +0200

pb-logging (0.4.2-3) wheezy; urgency=medium

  * Drop unused python-support from dependencies.

 -- Benjamin Drung <benjamin.drung@profitbricks.com>  Mon, 12 May 2014 14:55:42 +0200

pb-logging (0.4.2-2) wheezy; urgency=medium

  * Disable DH_VERBOSE.
  * Simplify debian/rules.
  * Rename source package to pb-logging.
  * Rename profitbricks-python-logging to python-pb-logging.
  * Bump Standard-Version to 3.9.5 (no changes required).
  * Add Homepage field pointing to gitweb.

 -- Benjamin Drung <benjamin.drung@profitbricks.com>  Mon, 12 May 2014 14:17:01 +0200

profitbricks-python-logging (0.4.2-1) wheezy; urgency=medium

  * Version bump

 -- Benjamin Drung <benjamin.drung@profitbricks.com>  Mon, 31 Mar 2014 11:05:44 +0200

profitbricks-python-logging (0.4.1-3) wheezy; urgency=low

  * Build for Wheezy

 -- Frank Brehm <frank.brehm@profitbricks.com>  Thu, 20 Mar 2014 14:29:07 +0100

profitbricks-python-logging (0.4.1-3) stable; urgency=low

  * Declared for stable

 -- Frank Brehm <frank.brehm@profitbricks.com>  Thu, 20 Mar 2014 14:24:45 +0100

profitbricks-python-logging (0.4.1-2) unstable; urgency=low

  * Added pb-logging.ebuild
  * Minimal changes in debian/rules

 -- Frank Brehm <frank.brehm@profitbricks.com>  Thu, 20 Mar 2014 14:16:13 +0100

profitbricks-python-logging (0.4.1-1) unstable; urgency=low

  * Modifications on debian build system for Python3

 -- Frank Brehm <frank.brehm@profitbricks.com>  Wed, 18 Dec 2013 19:32:56 +0100

profitbricks-python-logging (0.4.0-1) unstable; urgency=low

  * Declared for unstable
  * Converting for Python 3

 -- Frank Brehm <frank.brehm@profitbricks.com>  Wed, 18 Dec 2013 14:46:53 +0100

profitbricks-python-logging (0.3.4-1) wheezy; urgency=low

  * build for wheezy

 -- Thomas Möricke <thomas.moericke@profitbricks.com>  Wed, 19 Mar 2014 13:24:43 +0100

profitbricks-python-logging (0.3.4-1) stable; urgency=low

  * Bugfix in pb_logging/unix_handler.py
  * Added test/general.py and test/test_syslog.py
  * Added test methods test_logging_syslog() and test_unix_syslog() to
    test/test_syslog.py

 -- Frank Brehm <frank.brehm@profitbricks.com>  Tue, 12 Nov 2013 17:48:56 +0100

profitbricks-python-logging (0.3.3-2) stable; urgency=low

  * Declared for stable

 -- Frank Brehm <frank.brehm@profitbricks.com>  Mon, 05 Aug 2013 16:07:23 +0200

profitbricks-python-logging (0.3.3-1) unstable; urgency=low

  * Declared for unstable
  * Better handling on logging to UNIX sockets in
    pb_logging/syslog_handler.py
  * Refactored UnixSyslogHandler.

 -- Frank Brehm <frank.brehm@profitbricks.com>  Mon, 05 Aug 2013 15:57:35 +0200


pb_logging/colored.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
@summary: additional logging formatter for colored output via console
"""

# Standard modules
import logging
# import os.path
# import sys
import copy

# Third party modules

# Own modules

# import pb_provisioning.common

# from pb_provisioning.common import to_unicode_or_bust, to_utf8_or_bust

__version__ = '0.2.0'

# =============================================================================
# Color coding module variables and helper functions

COLOR_CODE = {
    'ENDC': 0,  # RESET COLOR
    'BOLD': 1,
    'UNDERLINE': 4,
    'BLINK': 5,
    'INVERT': 7,
    'CONCEALD': 8,
    'STRIKE': 9,
    'GREY30': 90,
    'GREY40': 2,
    'GREY65': 37,
    'GREY70': 97,
    'GREY20_BG': 40,
    'GREY33_BG': 100,
    'GREY80_BG': 47,
    'GREY93_BG': 107,
    'DARK_RED': 31,
    'RED': 91,
    'RED_BG': 41,
    'LIGHT_RED_BG': 101,
    'DARK_YELLOW': 33,
    'YELLOW': 93,
    'YELLOW_BG': 43,
    'LIGHT_YELLOW_BG': 103,
    'DARK_BLUE': 34,
    'BLUE': 94,
    'BLUE_BG': 44,
    'LIGHT_BLUE_BG': 104,
    'DARK_MAGENTA': 35,
    'PURPLE': 95,
    'MAGENTA_BG': 45,
    'LIGHT_PURPLE_BG': 105,
    'DARK_CYAN': 36,
    'AUQA': 96,
    'AQUA': 96,
    'CYAN_BG': 46,
    'LIGHT_AUQA_BG': 106,
    'LIGHT_AQUA_BG': 106,
    'DARK_GREEN': 32,
    'GREEN': 92,
    'GREEN_BG': 42,
    'LIGHT_GREEN_BG': 102,
    'BLACK': 30,
}


# -----------------------------------------------------------------------------
def termcode(num):
    """
    Output of an ANSII terminal code.
    """

    return('\033[%sm' % (num))


# -----------------------------------------------------------------------------
def colorstr(message, color):
    """
    Wrapper function to colorize the message.

    @param message: The message to colorize
    @type message: str
    @param color: The color to use, must be one of the keys of COLOR_CODE
    @type color: str

    @return: the colorized message
    @rtype: str

    """

    tcode = ''
    if isinstance(color, (list, tuple)):
        for clr in color:
            tcode += termcode(COLOR_CODE[clr])
    else:
        tcode = termcode(COLOR_CODE[color])

    return tcode + message + termcode(COLOR_CODE['ENDC'])


LOG = logging.getLogger(__name__)


# =============================================================================
class ColoredFormatter(logging.Formatter):
    """
    A variant of code found at:
    http://stackoverflow.com/questions/384076/how-can-i-make-the-python-logging-output-to-be-colored    # noqa
    """

    LEVEL_COLOR = {
        'DEBUG': None,
        'INFO': 'GREEN',
        'WARNING': 'YELLOW',
        'ERROR': ('BOLD', 'RED'),
        'CRITICAL': 'RED_BG',
    }

    # -------------------------------------------------------------------------
    def __init__(self, fmt=None, datefmt=None, dark=False):
        """
        Initialize the formatter with specified format strings.

        Initialize the formatter either with the specified format string, or a
        default. Allow for specialized date formatting with the optional
        datefmt argument (if omitted, you get the ISO8601 format).
        """

        logging.Formatter.__init__(self, fmt, datefmt)
        if dark:
            # changing the default colors to "dark" because the xterm plugin
            # for Jenkins cannot use bright colors
            # see: http://stackoverflow.com/a/28071761
            self.color_debug = 'DARK_CYAN'
            self.color_info = 'DARK_GREEN'
            self.color_warning = 'DARK_YELLOW'
            self.color_error = 'DARK_RED'

    # -----------------------------------------------------------
    @property
    def color_debug(self):
        """The color used to output debug messages."""
        return self.LEVEL_COLOR['DEBUG']

    @color_debug.setter
    def color_debug(self, value):
        self.LEVEL_COLOR['DEBUG'] = value

    # -----------------------------------------------------------
    @property
    def color_info(self):
        """The color used to output info messages."""
        return self.LEVEL_COLOR['INFO']

    @color_info.setter
    def color_info(self, value):
        self.LEVEL_COLOR['INFO'] = value

    # -----------------------------------------------------------
    @property
    def color_warning(self):
        """The color used to output warning messages."""
        return self.LEVEL_COLOR['WARNING']

    @color_warning.setter
    def color_warning(self, value):
        self.LEVEL_COLOR['WARNING'] = value

    # -----------------------------------------------------------
    @property
    def color_error(self):
        """The color used to output error messages."""
        return self.LEVEL_COLOR['ERROR']

    @color_error.setter
    def color_error(self, value):
        self.LEVEL_COLOR['ERROR'] = value

    # -----------------------------------------------------------
    @property
    def color_critical(self):
        """The color used to output critical messages."""
        return self.LEVEL_COLOR['CRITICAL']

    @color_critical.setter
    def color_critical(self, value):
        self.LEVEL_COLOR['CRITICAL'] = value

    # -------------------------------------------------------------------------
    def format(self, record):
        """
        Format the specified record as text.
        """

        record = copy.copy(record)
        levelname = record.levelname

        if levelname in self.LEVEL_COLOR:

            record.name = colorstr(record.name, 'BOLD')
            record.filename = colorstr(record.filename, 'BOLD')
            record.module = colorstr(record.module, 'BOLD')
            record.funcName = colorstr(record.funcName, 'BOLD')
            record.pathname = colorstr(record.pathname, 'BOLD')
            record.processName = colorstr(record.processName, 'BOLD')
            record.threadName = colorstr(record.threadName, 'BOLD')

            if self.LEVEL_COLOR[levelname] is not None:
                record.levelname = colorstr(
                    levelname, self.LEVEL_COLOR[levelname])
                record.msg = colorstr(record.msg, self.LEVEL_COLOR[levelname])

        return logging.Formatter.format(self, record)


# =============================================================================

if __name__ == "__main__":

    msg = 'Hallo Ballo!'

    for key in sorted(COLOR_CODE.keys()):
        print('%s: %s' % (key, colorstr(msg, key)))

    print("")
    print((colorstr('Bold Red', ('BOLD', 'RED'))))
    print((colorstr('Normal Red', 'RED')))

    print("")

    LOG.setLevel(logging.DEBUG)
    console = logging.StreamHandler()
    console.setFormatter(
        ColoredFormatter('%(name)s: %(message)s (%(filename)s:%(lineno)d)'))
    LOG.addHandler(console)

    fh = logging.FileHandler('/tmp/test.log', 'w')
    fh.setFormatter(logging.Formatter('%(name)s: %(message)s'))
    LOG.addHandler(fh)

    LOG.debug('debug')
    LOG.info('info')
    LOG.warning('Warning')
    LOG.error('ERROR')
    LOG.critical('CRITICAL!!!')


# =============================================================================

# vim: tabstop=4 expandtab shiftwidth=4 softtabstop=4
